### PR TITLE
Adding LLVM14 - Missing config but says supported in Error message.

### PR DESCRIPTION
**Pull request by Pix-xiP**
As it says on the tin, the build_odin.sh script says 14 is supported in the list of errors, however there is no check in the script to set it. Just popped it in and tested on Arch with the latest build and it worked with 14.

### CODE REVIEW

**Comment by jcmdln** [comment]
This has come up a lot so I agree it should just be handled. LLVM-14 is not officially supported so if there are any issues discovered with Odin + LLVM14 they are not likely to be patched, which is why this was made an explicit opt-in.

Personally I would like to see LLVM14 added to the `# fallback` section so that all officially supported versions of LLVM are tried first. Also the FreeBSD commands as they are are wrong and could be corrected:

```diff
diff --git a/build_odin.sh b/build_odin.sh
index 2a2505c9..ea6ca328 100755
--- a/build_odin.sh
+++ b/build_odin.sh
@@ -31,14 +31,16 @@ if [ -z "$LLVM_CONFIG" ]; then
        elif [ -n "$(command -v llvm-config-12)" ]; then LLVM_CONFIG="llvm-config-12"
        elif [ -n "$(command -v llvm-config-11)" ]; then LLVM_CONFIG="llvm-config-11"
        # freebsd
-       elif [ -n "$(command -v llvm-config17)" ]; then  LLVM_CONFIG="llvm-config-17"
-       elif [ -n "$(command -v llvm-config13)" ]; then  LLVM_CONFIG="llvm-config-13"
-       elif [ -n "$(command -v llvm-config12)" ]; then  LLVM_CONFIG="llvm-config-12"
-       elif [ -n "$(command -v llvm-config11)" ]; then  LLVM_CONFIG="llvm-config-11"
+       elif [ -n "$(command -v llvm-config17)" ]; then  LLVM_CONFIG="llvm-config17"
+       elif [ -n "$(command -v llvm-config13)" ]; then  LLVM_CONFIG="llvm-config13"
+       elif [ -n "$(command -v llvm-config12)" ]; then  LLVM_CONFIG="llvm-config12"
+       elif [ -n "$(command -v llvm-config11)" ]; then  LLVM_CONFIG="llvm-config11"
        # fallback
+       elif [ -n "$(command -v llvm-config-14)" ]; then LLVM_CONFIG="llvm-config-14"
+       elif [ -n "$(command -v llvm-config14)" ]; then  LLVM_CONFIG="llvm-config14"
        elif [ -n "$(command -v llvm-config)" ]; then LLVM_CONFIG="llvm-config"
        else
-               error "No llvm-config command found. Set LLVM_CONFIG to proceed."
+               error "No supported llvm-config command found. Set LLVM_CONFIG to proceed."
        fi
 fi
 ```

**Comment by jcmdln** [comment]
To clarify, the ordering is important so that when Odin cuts releases for Linux it uses an officially supported version of LLVM. If LLVM14 is added above LLVM13 then the pipeline would attempt to use LLVM14 which is bundled with GitHub's `actions/runner-images`.

If it's instead added to the `# fallback` section, the Odin pipeline will continue to cut releases as expected and users who have distributions that only provide LLVM14 will be able to build as they currently do, as this version restriction was never enforced in the first place.

**Comment by Pix-xiP** [comment]
Makes sense to me, just figured it should probably exist in there if the message is gonna tell people about it, I pulled 14 cause it mentioned,  builds fine on Arch, but I dunno about other system.

**Comment by jcmdln** [comment]
I agree. Maybe this is a good opportunity to add some kind of warning message?

```
diff --git a/build_odin.sh b/build_odin.sh
index 2a2505c9..025e51cd 100755
--- a/build_odin.sh
+++ b/build_odin.sh
@@ -31,14 +31,20 @@ if [ -z "$LLVM_CONFIG" ]; then
        elif [ -n "$(command -v llvm-config-12)" ]; then LLVM_CONFIG="llvm-config-12"
        elif [ -n "$(command -v llvm-config-11)" ]; then LLVM_CONFIG="llvm-config-11"
        # freebsd
-       elif [ -n "$(command -v llvm-config17)" ]; then  LLVM_CONFIG="llvm-config-17"
-       elif [ -n "$(command -v llvm-config13)" ]; then  LLVM_CONFIG="llvm-config-13"
-       elif [ -n "$(command -v llvm-config12)" ]; then  LLVM_CONFIG="llvm-config-12"
-       elif [ -n "$(command -v llvm-config11)" ]; then  LLVM_CONFIG="llvm-config-11"
+       elif [ -n "$(command -v llvm-config17)" ]; then  LLVM_CONFIG="llvm-config17"
+       elif [ -n "$(command -v llvm-config13)" ]; then  LLVM_CONFIG="llvm-config13"
+       elif [ -n "$(command -v llvm-config12)" ]; then  LLVM_CONFIG="llvm-config12"
+       elif [ -n "$(command -v llvm-config11)" ]; then  LLVM_CONFIG="llvm-config11"
        # fallback
+       elif [ -n "$(command -v llvm-config-14)" ]; then
+               echo "WARN: Odin does not officially support LLVM14"
+               LLVM_CONFIG="llvm-config-14"
+       elif [ -n "$(command -v llvm-config14)" ]; then
+               echo "WARN: Odin does not officially support LLVM14"
+               LLVM_CONFIG="llvm-config14"
        elif [ -n "$(command -v llvm-config)" ]; then LLVM_CONFIG="llvm-config"
        else
-               error "No llvm-config command found. Set LLVM_CONFIG to proceed."
+               error "No supported llvm-config command found. Set LLVM_CONFIG to proceed."
        fi
 fi
 ```

**Comment by Pix-xiP** [comment]
Not opposed to that either, more information is better in this case imo. 

Looks good to me otherwise unless you want me to change anything else

File: build_odin.sh [hunk 1]
```diff
@@ -27,11 +27,13 @@ error() {
 if [ -z "$LLVM_CONFIG" ]; then
 	# darwin, linux, openbsd
 	if   [ -n "$(command -v llvm-config-17)" ]; then LLVM_CONFIG="llvm-config-17"
+	elif [ -n "$(command -v llvm-config-14)" ]; then LLVM_CONFIG="llvm-config-14"
```
Comment (jcmdln) [inline build_odin.sh:30]: Move this below the `llvm-config-11` condition in this section

File: build_odin.sh [hunk 2]
```diff
@@ -27,11 +27,13 @@ error() {
...
 	elif [ -n "$(command -v llvm-config-13)" ]; then LLVM_CONFIG="llvm-config-13"
 	elif [ -n "$(command -v llvm-config-12)" ]; then LLVM_CONFIG="llvm-config-12"
 	elif [ -n "$(command -v llvm-config-11)" ]; then LLVM_CONFIG="llvm-config-11"
 	# freebsd
 	elif [ -n "$(command -v llvm-config17)" ]; then  LLVM_CONFIG="llvm-config-17"
+	elif [ -n "$(command -v llvm-config14)" ]; then  LLVM_CONFIG="llvm-config-14"
```
Comment (jcmdln) [inline build_odin.sh:36]: Move this below the `llvm-config11` condition in this section

**Review by jcmdln — CHANGES_REQUESTED** [review]
At minimum, these are the changes that I think should happen.

Without these changes, the Odin pipeline will implicitly use LLVM-14 which is not officially supported, but apparently works. This is visible in the pipeline associated with the current commit: https://github.com/odin-lang/Odin/actions/runs/6642566987/job/18047619005?pr=2893#step:4:5

**Comment by Pix-xiP** [comment]
Mibad, I though you had made edits when I read this before, didn't realise you were just pushing ideas.

I'll add the warning version and see if that vibes

File: build_odin.sh [hunk 3]
```diff
@@ -36,7 +36,12 @@ if [ -z "$LLVM_CONFIG" ]; then
 	elif [ -n "$(command -v llvm-config12)" ]; then  LLVM_CONFIG="llvm-config-12"
 	elif [ -n "$(command -v llvm-config11)" ]; then  LLVM_CONFIG="llvm-config-11"
 	# fallback
-	elif [ -n "$(command -v llvm-config)" ]; then LLVM_CONFIG="llvm-config"
+  elif [ -n "$(command -v llvm-config-14)" ]; then
+    echo "WARN: Odin does not officially support LLVM14"
+    LLVM_CONFIG="llvm-config-14"
+  elif [ -n "$(command -v llvm-config14)" ]; then
+    echo "WARN: Odin does not officially support LLVM14"
+    LLVM_CONFIG="llvm-config14"
```
Comment (jcmdln) [inline build_odin.sh:44]: Please use tabs instead of spaces

**Comment by jcmdln** [comment]
Thanks for making the changes @Pix-xiP! Sorry I should have explained that those were just ideas I had. Last thing is a minor style issue which I've left a comment for. After that's resolved I think this this is ready to be merged though we'll have to wait for a maintainer to do so.

The pipeline is failing on the `HEAD` of the `master` branch, so ignore the errors regarding it for now as they are unrelated to this PR.

File: build_odin.sh (hunk 3 above)
Comment (Pix-xiP) [reply build_odin.sh]: Done!

**Comment by Pix-xiP** [comment]
Cool, hopefully someone with the big guns gets a free moment to peek this then. 

My bad on the spaces, it didn't even register that it might use tabs.

**Comment by flysand7** [comment]
Related: #3013

**Comment by laytan** [comment]
Looks like this has been added at some point somewhere else

**Pull request closed without merging**